src/client.rs

```diff
@@ -0,0 +1,162 @@
+    pub audiences: Audiences,
+    /// `Resend` APIs for `METHOD /audiences/:id/contacts` endpoints.
+    pub contacts: Contacts,
+    /// `Resend` APIs for `METHOD /domains` endpoints.
+    pub domains: Domains,
+}
+
+#[derive(Clone)]
+pub struct Config {
+    user_agent: String,
+    api_key: String,
+    pub(crate) base_url: Url,
+    pub(crate) client: ReqwestClient,
+}
+
+impl Config {
+    /// Constructs a new [`RequestBuilder`] with parameters.
+    pub fn build(&self, method: Method, path: &str) -> RequestBuilder {
+        let path = self
+            .base_url
+            .join(path)
+            .expect("should a valid API endpoint");
+
+        self.client
+            .request(method, path)
+            .bearer_auth(self.api_key.as_str())
+            .header(USER_AGENT, self.user_agent.as_str())
+    }
+}
+
+impl fmt::Debug for Config {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        // Don't output API key.
+        f.debug_struct("Client")
+            .field("api_key", &"re_*********")
+            .field("user_agent", &self.user_agent.as_str())
+            .field("base_url", &self.base_url.as_str())
+            // .field("client", &self.client)
+            .finish()
+    }
+}
+
+impl Client {
+    /// Creates a new [`Resend`] client.
+    ///
+    /// ### Panics
+    ///
+    /// Panics if the environment variable `RESEND_BASE_URL` is present but is not a valid URL.
+    ///
+    /// [`Resend`]: https://resend.com
+    pub fn new(api_key: &str) -> Self {
+        Self::with_client(api_key, ReqwestClient::default())
+    }
+
+    /// Creates a new [`Resend`] client with a provided [`reqwest::Client`].
+    ///
+    /// ### Panics
+    ///
+    /// Panics if the environment variable `RESEND_BASE_URL` is present but is not a valid URL.
+    ///
+    /// [`Resend`]: https://resend.com
+    /// [`reqwest::Client`]: ReqwestClient
+    pub fn with_client(api_key: &str, client: ReqwestClient) -> Self {
+        let env_base_url = env::var("RESEND_BASE_URL")
+            .map_or_else(
+                |_| Url::parse("https://api.resend.com"),
+                |env_var| Url::parse(env_var.as_str()),
+            )
+            .expect("env variable `RESEND_BASE_URL` should be a valid URL");
+
+        let env_user_agent = env::var("RESEND_USER_AGENT").unwrap_or_else(|_| {
+            format!("{}/{}", env!("CARGO_PKG_NAME"), env!("CARGO_PKG_VERSION"))
+        });
+
+        let inner = Arc::new(Config {
+            user_agent: env_user_agent,
+            api_key: api_key.to_owned(),
+            base_url: env_base_url,
+            client,
+        });
+
+        Self {
+            api_keys: ApiKeys(inner.clone()),
+            audiences: Audiences(inner.clone()),
+            contacts: Contacts(inner.clone()),
+            domains: Domains(inner.clone()),
+            emails: Emails(inner),
+        }
+    }
+
+    /// Returns the used `User-Agent` header value.
+    #[must_use]
```

Comment:
What is the rationale behind these? I've never used them myself. I know they raise a warning when you don't use the returned value but why add it in these methods specifically? Asking out of curiosity, not really a review comment 😅

Reply:
Because these methods don't have any side effects, not using the return value is likely to result in a logical error. The same applies to responses from retrieve and list endpoints.

Reply:
Makes sense, thanks!